### PR TITLE
Replay should respect order of register_ticks with respect to blockhashes

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
We register multiple blockhashes for the same slot if that slot's block contains ticks from other blocks
(virtual ticks). Because `process_entries_with_callback` doesn't respect ledger ordering of ticks, this can cause validators to have different replay behavior

#### Summary of Changes
Only register blockhash when tick_height == max_tick_height

Fixes #

### CODE REVIEW

**Comment by garious** [comment]
I don't understand why the problem you've described is a problem. By ignoring virtual ticks, the blockhash queue can represent a hash chain longer than 2 minutes. If you allow that, it becomes an attack vector to DoS leaders until you're ready to send in your otherwise too old transaction.

**Comment by mvines** [comment]
The other way to do it is to make process_entries_with_callback() respect the actual ledger ordering of ticks, so that transactions can safely refer to the blockhashes of virtual blocks

**Comment by carllin** [comment]
@garious, good point, I think @mvines is right

File: ledger/src/blocktree_processor.rs [hunk 1]
```diff
@@ -117,7 +117,18 @@ fn process_entries_with_callback(
         if entry.is_tick() {
             // if its a tick, save it for later
             tick_hashes.push(entry.hash);
-            continue;
+            if bank.is_block_boundary(bank.tick_height() + tick_hashes.len() as u64) {
+                // if its a tick that will cause a new blockhash to be created,
```
Comment (garious) [inline ledger/src/blocktree_processor.rs]: It's

Comment (carllin) [reply ledger/src/blocktree_processor.rs]: Oops yup, will also write a test after I get back from dinner

**Comment by carllin** [comment]
@sakridge this is the fix for the issue we discussed

**Comment by sakridge** [comment]
> @sakridge this is the fix for the issue we discussed, hopefully replay perf isn't much worse

I see. This should be OK since we are only syncing on the slot boundary.

File: ledger/src/blocktree_processor.rs [hunk 2]
```diff
@@ -115,8 +115,18 @@ fn process_entries_with_callback(
...
-            // if its a tick, save it for later
+            // If it's a tick, save it for later
             tick_hashes.push(entry.hash);
+            if bank.is_block_boundary(bank.tick_height() + tick_hashes.len() as u64) {
+                // If it's a tick that will cause a new blockhash to be created,
+                // execute the group and register the tick
+                execute_batches(bank, &batches, entry_callback)?;
+                batches.clear();
+                for hash in &tick_hashes {
+                    bank.register_tick(hash);
```
Comment (rob-solana) [inline ledger/src/blocktree_processor.rs]: with all these a-priori checks, does the bank need to see these hashes anymore?  how about just bank.increment_ticks() + bank.register_blockhash()?

**Pull request merged**